File: ts/ai-agent-api/ai/ai.ts
import { api, APIError } from "encore.dev/api";
import { secret } from "encore.dev/config";

const anthropicAPIKey = secret("AnthropicAPIKey");

export interface Message {
  role: "user" | "assistant";
  content: string;
}

interface CompleteRequest {
  messages: Message[];
}

interface CompleteResponse {
  content: string;
}

// Internal API for generating AI completions using Claude.
export const complete = api(
  { expose: false, method: "POST", path: "/ai/complete" },
  async ({ messages }: CompleteRequest): Promise<CompleteResponse> => {
    const resp = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-api-key": anthropicAPIKey(),
        "anthropic-version": "2023-06-01",
      },
      body: JSON.stringify({
        model: "claude-sonnet-4-6",
        max_tokens: 1024,
        messages: messages.map((m) => ({
          role: m.role,
          content: m.content,
        })),
      }),
    });

    if (!resp.ok) {
      const body = await resp.text();
      throw APIError.internal(`AI request failed: ${resp.status} ${body}`);
    }

    const data = (await resp.json()) as {
      content: { type: string; text: string }[];
    };

    const text = data.content
      .filter((c) => c.type === "text")
      .map((c) => c.text)
      .join("");

    return { content: text };
  },
);

File: ai-agent-api/ai/ai.go
// Service ai provides AI completion capabilities using the Anthropic Claude API.
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

var secrets struct {
	AnthropicAPIKey string
}

// Message represents a single message in a conversation.
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// CompleteRequest is the request to generate an AI completion.
type CompleteRequest struct {
	Messages []Message `json:"messages"`
}

// CompleteResponse is the response from the AI completion.
type CompleteResponse struct {
	Content string `json:"content"`
}

// Complete generates an AI completion using the Anthropic Claude API.
//
//encore:api private method=POST path=/ai/complete
func Complete(ctx context.Context, req *CompleteRequest) (*CompleteResponse, error) {
	body := map[string]any{
		"model":      "claude-sonnet-4-6",
		"max_tokens": 1024,
		"messages":   req.Messages,
	}
	bodyJSON, err := json.Marshal(body)
	if err != nil {
		return nil, fmt.Errorf("marshal request: %v", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", "https://api.anthropic.com/v1/messages", bytes.NewReader(bodyJSON))
	if err != nil {
		return nil, fmt.Errorf("create request: %v", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("x-api-key", secrets.AnthropicAPIKey)
	httpReq.Header.Set("anthropic-version", "2023-06-01")

	resp, err := http.DefaultClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("call anthropic: %v", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response: %v", err)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("anthropic API error (%d): %s", resp.StatusCode, respBody)
	}

	var result struct {
		Content []struct {
			Type string `json:"type"`
			Text string `json:"text"`
		} `json:"content"`
	}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("unmarshal response: %v", err)
	}

	var text string
	for _, block := range result.Content {
		if block.Type == "text" {
			text += block.Text
		}
	}

	return &CompleteResponse{Content: text}, nil
}
